### PR TITLE
cf-252 Change wait_until_* methods to not remove from list.

### PR DESCRIPTION
**Pull request by plomakin**
- Get rid of removing items from list which it iterates on, this may not always work as expected;
 - All the {{wait_until_.*}} methods have very similar if not the same implementation, thus need to be generalized and moved to a single method.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.generate_load
devlab/generate_load.py:591: [E1101(no-member), Prerequisites.create_security_group] Instance of 'dict' has no 'id' member
devlab/generate_load.py:665: [E1101(no-member), Prerequisites.create_cinder_volumes] Instance of 'dict' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.generate_load
devlab/generate_load.py:591: [E1101(no-member), Prerequisites.create_security_group] Instance of 'dict' has no 'id' member
devlab/generate_load.py:665: [E1101(no-member), Prerequisites.create_cinder_volumes] Instance of 'dict' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by SVilgelm** [comment]
:-1: Please take a look at this line: https://github.com/MirantisWorkloadMobility/CloudFerry/pull/865/files#diff-b6fd6ed2cd7b743411a1944f8095e451R220

**Comment by SVilgelm** [comment]
-2: I don't agree with you solution, you have to rewrite all those functions and also the retry_until_resources_created decorator too. My suggestions is:
* to implement the function that will take a list of objects and a function to check the status. 
* to implement a checking function for each type of objects instead of current wait_until... functions

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.generate_load
devlab/generate_load.py:562: [E1101(no-member), 
Prerequisites.create_security_group] Instance of 'dict' has no 'id' member
devlab/generate_load.py:625: [E1101(no-member), 
Prerequisites.create_cinder_volumes] Instance of 'dict' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.generate_load
devlab/generate_load.py:562: [E1101(no-member), 
Prerequisites.create_security_group] Instance of 'dict' has no 'id' member
devlab/generate_load.py:625: [E1101(no-member), 
Prerequisites.create_cinder_volumes] Instance of 'dict' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: devlab/tests/base.py [hunk 1]
```diff
@@ -246,10 +247,50 @@ def user_has_not_primary_tenants(self, user_name):
...
+            msg = 'Snapshot with id {0} has become in error state'
+            raise RuntimeError(msg.format(snp_id))
+        return False
+
+    @staticmethod
+    def wait_until_objects_created(obj_list, check_func, timeout):
+        waiting = 0
+        delay = 1
+        while waiting < timeout:
+            time.sleep(delay)
```
Comment (SVilgelm) [inline devlab/tests/base.py]: nooo, you did nothing and already want to wait, why? Move time.sleep to the end of this block

File: devlab/generate_load.py [hunk 2]
```diff
@@ -217,7 +193,8 @@ def _get_body_for_image_creating(_image):
             img = self.glanceclient.images.create(
                 **_get_body_for_image_creating(image))
             img_ids.append(img.id)
-        wait_until_images_created(img_ids[:])
+        self.wait_until_objects_created(img_ids[:], self.check_image_state,
```
Comment (SVilgelm) [inline devlab/generate_load.py]: I thin you don't need to make a copy of list here

Comment (SVilgelm) [inline devlab/generate_load.py]: Or you have to make copy of list in other places too

File: devlab/tests/base.py (hunk 1 above)
Comment (plomakin) [inline devlab/tests/base.py]: sure

File: devlab/generate_load.py (hunk 2 above)
Comment (SVilgelm) [inline devlab/generate_load.py]: No, This is correct code, you don't need to modify. In the other places we are not using the list of id after calling self.wait_until_objects_created function.

Comment (plomakin) [inline devlab/generate_load.py]: I will re-use the same image list only in this code block. In other cases i don't need to save lists like this.

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.generate_load
devlab/generate_load.py:562: [E1101(no-member), 
Prerequisites.create_security_group] Instance of 'dict' has no 'id' member
devlab/generate_load.py:625: [E1101(no-member), 
Prerequisites.create_cinder_volumes] Instance of 'dict' has no 'id' member

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by plomakin** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module devlab.generate_load
devlab/generate_load.py:562: [E1101(no-member), 
Prerequisites.create_security_group] Instance of 'dict' has no 'id' member
devlab/generate_load.py:625: [E1101(no-member), 
Prerequisites.create_cinder_volumes] Instance of 'dict' has no 'id' member

File: devlab/tests/base.py [hunk 3]
```diff
@@ -259,8 +299,7 @@ def tenant_exists(self, tenant_name=None, tenant_id=None):
                 self.keystoneclient.tenants.find(id=tenant_id)
         except ks_exceptions.NotFound:
             return False
-        else:
-            return True
+        return True
```
Comment (ntymtsiv) [inline devlab/tests/base.py]: What's the difference?

**Comment by ntymtsiv** [comment]
+1

File: devlab/tests/base.py (hunk 3 above)
Comment (plomakin) [inline devlab/tests/base.py]: no differences, just making code more compact ;-)

**Comment by SVilgelm** [comment]
:+1: can be merged

**Comment by roman-verchikov** [comment]
+1, can be merged

**Pull request merged**